GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,171 @@
+            loggingFunction(
+                metadata,
+                string.Format(
+                    messageTemplate,
+                    numberOfCertificates,
+                    string.Join(
+                        Environment.NewLine,
+                        certificates.Select(x => x.Subject))));
+        }
+
+        private X509Certificate2 GetCertificateFromFile(ITracer tracer, EventMetadata metadata, string certificatePassword, bool onlyLoadValidCertificateFromStore)
+        {
+            if (File.Exists(this.CertificatePathOrSubjectCommonName))
+            {
+                try
+                {
+                    X509Certificate2 cert = new X509Certificate2(this.CertificatePathOrSubjectCommonName, certificatePassword);
+                    if (onlyLoadValidCertificateFromStore && cert != null && !cert.Verify())
+                    {
+                        tracer.RelatedWarning(metadata, "Certficate was found, but is invalid.");
+                        return null;
+                    }
+
+                    return cert;
+                }
+                catch (CryptographicException cryptEx)
+                {
+                    metadata.Add("Exception", cryptEx);
+                    tracer.RelatedError(metadata, "Error, while loading certificate from disk");
+                    return null;
+                }
+            }
+
+            return null;
+        }
+
+        private X509Certificate2 GetCertificateFromStore(ITracer tracer, EventMetadata metadata, bool onlyLoadValidCertificateFromStore)
+        {
+            X509Store store = null;
+            try
+            {
+                store = new X509Store();
+                store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+
+                X509Certificate2Collection findResults = store.Certificates.Find(X509FindType.FindBySubjectName, this.CertificatePathOrSubjectCommonName, onlyLoadValidCertificateFromStore);
+                if (findResults?.Count > 0)
+                {
+                    LogCertificateCounts(tracer, metadata, findResults.OfType<X509Certificate2>(), "Found {0} certificates by provided name. Matching DNs: {1}");
+
+                    X509Certificate2[] certsWithMatchingCns = findResults
+                        .OfType<X509Certificate2>()
+                        .Where(x => x.HasPrivateKey && Regex.IsMatch(x.Subject, string.Format("(^|,\\s?)CN={0}(,|$)", this.CertificatePathOrSubjectCommonName))) // We only want certificates, that have private keys, as we need them. We also want a complete CN match
+                        .OrderByDescending(x => x.Verify()) // Ordering by validity in a descending order will bring valid certificates to the beginning
+                        .ThenBy(x => x.NotBefore) // We take the one, that was issued earliest, first
+                        .ThenByDescending(x => x.NotAfter) // We then take the one, that is valid for the longest period
+                        .ToArray();

```

Comment:
You can remove the call to `ToArray` and eliminate enumerating the collection multiple times.

Reply:
But `ToArray` is giving us precisely that - preventing from enumerating the whole collection multiple times. If we remove the `ToArray` call, we'll evaluate the whole expression three times:
1. When getting the number of results for logging
2. When selecting the subject names for logging
3. On the `return` statement`, when selecting the `FirstOrDefault`

When we do `ToArray`, we only operate on the resulting array in all the three cases, without any ordering, or filtering.

Reply:
Thanks I see that now.